### PR TITLE
2509 - Data submission email

### LINKED ISSUE #2509: As a data analyst, I need to know when my data has been processed with or w/o errors
**Description:**
Generate email to STT data analysts for error reports on submitted datafiles. this email will replace the current data submission email that STT users currently receive. 

**Acceptance Criteria:** 
- [ ] Aligned with design spec ("Near Term Lower-lift" version):
   - [ ] Data analysts receive email notification that their data is processed
   - [ ] Email subject leads with whether or not their data file had errors
   - [ ] Email table contains columns for section, submitted by, and error report (Error Report Available/No Errors being possible values)
   - [ ] Email body contains copy on where to access the error report in TDP and where to go for more information in the knowledge center
- [ ] Replaces existing notifications for data file submission
- [ ] Testing Checklist has been run and all tests pass
- [ ] README is updated, if necessary

**Tasks:**

- [x] Create new email template [here](https://github.com/raft-tech/TANF-app/tree/develop/tdrs-backend/tdpservice/email/templates)
- [x] Find conditional/trigger for when error reports are ready
- [x] Test null values for all variables
- [x] Remove old email template for data submission**
- [ ] Run Testing Checklist and confirm all tests pass

**Notes:**
- Pull in after 1) Parsing Enhancements and 2) Error Messaging epics are complete.

**Supporting Documentation:**
- [Design spec (HTML)](https://reitermb.github.io/TDP-HTML-Email-Demo-Environment/error-report-available-low-lift.html)

Screenshot w/ replacement tags
image

Screenshot w/ example values
image

**Open Questions:**

- How can this be made as a stepping stone to ultimately #2248 ?
- Do we indeed want to replace the old email template for data submission even for minutes delay?

### PR DESCRIPTION
**Pull request by jtimpe**
## Summary of Changes
Pull request closes #2509. Updates the `send_data_submitted_email` to
* include more information in the template
* send once parsing completes, rather when file upload completes
* distinguish "processed with errors" and "processed without errors" in the subject and preview line

## How to Test

1. Update `SENDGRID_API_KEY` in your `.env` to a valid api key (i can provide one if you reach out). ensure `SENDGRID_SANDBOX_MODE_IN_DEBUG` remains `False`.
1. Run
   ```
   cd tdrs-backend && docker-compose up
   cd tdrs-frontend && docker-compose up --build
   ```
1. Set your account type to "Data Analyst"
1. Upload a data file (both containing errors and containing no errors)
1. Inspect the emails you receive

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
- [x] Aligned with design spec ("Near Term Lower-lift" version):
   - [x] Data analysts receive email notification that their data is processed
   - [x] Email subject leads with whether or not their data file had errors
   - [x] Email table contains columns for section, submitted by, and error report (Error Report Available/No Errors being possible values)
   - [x] Email body contains copy on where to access the error report in TDP and where to go for more information in the knowledge center
- [x] Replaces existing notifications for data file submission
- [x] Testing Checklist has been run and all tests pass
+ [x] **`lfrohlich`** and/or **`adpennington`**  confirmed that ACs are met.

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [x] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [x] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [x] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?

### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

### CODE REVIEW

File: tdrs-backend/tdpservice/email/templates/data-submitted.html [hunk 1]
```diff
@@ -1,39 +1,61 @@
...
+        <a href="https://tdp-project-updates.app.cloud.gov/knowledge-center/view-submission-history.html" target="_blank">
+            Read more about using Submission History
+        </a>
+    </p>
+
+
+    <p>
+        <a
+            class="button"
+            href="https://tdp-frontend-a11y.app.cloud.gov/"
```
Comment (elipe17) [inline tdrs-backend/tdpservice/email/templates/data-submitted.html]: Should this link be parametrized?

**Review by elipe17 — COMMENTED** [review]
Question about the link in the template.

File: tdrs-backend/tdpservice/email/templates/data-submitted.html (hunk 1 above)
Comment (jtimpe) [reply tdrs-backend/tdpservice/email/templates/data-submitted.html]: yes it should! good callout

**Review by elipe17 — APPROVED** [review]
Love the test coverage!

**Review by raftmsohani — APPROVED** [review]
LGTM, tested on local! Great work!

**Review by atrimpe — COMMENTED** [review]
LGTM

**Review by atrimpe — APPROVED** [review]
LGTM

**Review by reitermb — APPROVED** [review]
Emails are coming through and looking good 🚀 

image

**Comment by ADPennington** [comment]
> Emails are coming through and looking good 🚀
> 
> image

@reitermb could you remind me -- is there an adjacent UX ticket that updates the [e-mail notifications page](https://tdp-project-updates.app.cloud.gov/knowledge-center/about-email-notifications.html) related to this change? I'm also curious if we have prepared language about when these will and wont get delivered? (e.g. not delivered if files are still in pending status)

**Review by ADPennington — COMMENTED** [review]
@jtimpe @reitermb this is in good shape but 2 nits:

- the file type is truncated for TANF and Tribal TANF. If its easy to resolve this in this ticket, please implement. 
- I was able to confirm that files that get stuck in pending state do not result in a data submission email. I resubmitted and waited 4 minutes for it to work through the processing again and can confirm that the email is delivered after processing. there were 27K parsing errors detected, for reference. this should be addressed in UX ticket for KC. 

----

- when the file has errors:
  - TANF: the `file type` is TAN but for better clarity, should be TANF
tanf_errors_email







  - SSP

ssp_errors_email


  - Tribal TANF - the `file type` is Tribal TAN but for better clarity, should be Tribal TANF

  - tribal_errors_email

- when the file does not have errors:
  - TANF

tanf_noerrors_email

  - SSP
ssp_noerrors_email

  - Tribal TANF
tribal_noerrors_email


- when the file is too mangled and stays in pending state...
  - if the file gets stuck in pending while processing, an email is not delivered
  - after resubmitting a file that got stuck  with 27K parsing errors, email was delivered (approx 4 minutes later) 
  - tanf_pending_email
- Other:
    - I've noticed latency when 3 or more very small files are submitted at once. The data submissions are received in some cases. submission notification on-screen doesnt appear, but emails do but usually only for a subset of files submitted. i think this is a pre-existing issue so need a bug ticket.

File: tdrs-backend/tdpservice/email/helpers/data_file.py [hunk 2]
```diff
@@ -1,25 +1,71 @@
...
+        'object_repr': f"Uploaded data file for quarter: {datafile.fiscal_year}"
+    }
+
+    template_path = None
+    subject = None
+    text_message = None
+
+    section_name = datafile.section
+    prog_type = get_prog_from_section(section_name)
+    file_type = f'{prog_type}F' if prog_type != 'SSP' else prog_type
```
Comment (ADPennington) [inline tdrs-backend/tdpservice/email/helpers/data_file.py:29]: @jtimpe thank you for this!

tanf_spelling

**Review by ADPennington — APPROVED** [review]
lgtm 🚀 

notes for future ticket tracking:
- @reitermb we need to data submission email and description updated in the Knowledge Center and not sure which ticket is tracking this work
- @ADPennington to generate a bug ticket to track testing submitting 3+ files at a time.

**Pull request merged**